src/main/java/com/google/step/datamanager/DealManager.java

```diff
@@ -17,4 +18,16 @@ public Deal createDeal(
   public Deal updateDeal(Deal deal);
 
   public void deleteDeal(long id);
+
+  public List<Deal> getTrendingDeals();
+
+  public List<Deal> getDealsPublishedByFollowedUsers(long userId);
+
+  public List<Deal> getDealsPublishedByFollowedRestaurants(long userId);
+
+  public List<Deal> getDealsPublishedByFollowedTags(long userId);
```

Comment:
Will you need to use the follow manager for this? I don't think it's a good idea to use a data manager inside another one. A data manager should only be concerned with it's own type of data, deal in this case. DealManager is just an abstraction over the deal entity data store. I would rather suggest doing this -
```suggestion
  public List<Deal> getDealsPublishedByUsers(List<Long> userIds);

  public List<Deal> getDealsPublishedByRestaurants(List<Long> restaurantIds);

  public List<Deal> getDealsPublishedWithTags(List<Long> tagIds);
```
The caller should first call the follow manager to get the list of followed users / restaurants / tags and then pass their ids here.

Reply:
So i should call the follow manager first from the servlet?

Reply:
For getDealsPublishedWithTags i will have to call dealTagManager to getDealIdsWithTag in the deal manager so will that be an issue? and I will also need to call voteManager to getVotes() for each deal

Reply:
you can consider moving some of the logic from dealManager to the home page servlets, then dealManager would not need to know so much. Eg, HomeServlet will call FollowManager to see what dealIds it needs, then HomeServlet will call dealManager for the relevant deals with those IDs.

Reply:
+1. Yes, I think servlet is a better place to put this logic. 
I think of this as a layered architecture. There is (lower level) database layer which is managers in our case and there's (higher level) application layer which is the servlets. You should avoid communication within the same layer, as in, never use a manager inside another manager, and a servlet inside another servlet. This ensures that input always flows from application layer to db layer and output the other way. Makes it much easier to debug issues when application grows big.

Reply:
i can change follow manager to the servlet but im not sure how to change the location of votemanager and dealtagmanager, since votemanager is needed to calculate the hot score as well, its in #61 , and i think its weird to put the calculation in the servlet. For dealTagManager, i need it in the deal manager to retrieve the deals associated with the list of tagIds

Reply:
Why is it weird to put the calculation in the servlet? Or maybe you can make a class `HotScoreCalculator` which can calculate the hotscore by passing in `numVotes`?